File: app/src/lib/server/fs/daily.ts
import { createNote, noteExists, readNote, getVaultPath } from "./vault.js";
import { readFile } from "node:fs/promises";
import { resolve } from "node:path";

function formatDate(date: Date): string {
  return date.toISOString().slice(0, 10);
}

function formatDateJa(date: Date): string {
  return date.toLocaleDateString("ja-JP", {
    year: "numeric",
    month: "long",
    day: "numeric",
    weekday: "long",
  });
}

async function getTemplate(): Promise<string> {
  const vaultPath = await getVaultPath();
  const configPath = resolve(vaultPath, ".prometheus/config.json");

  try {
    const config = JSON.parse(await readFile(configPath, "utf-8"));
    if (config.dailyNoteTemplate) {
      return config.dailyNoteTemplate;
    }
  } catch {
    // Use default template
  }

  return `---
title: "{{date}}"
tags: [daily]
---

# {{dateJa}}

## Tasks

- [ ]

## Notes

`;
}

function applyTemplate(template: string, date: Date): string {
  return template
    .replace(/\{\{date\}\}/g, formatDate(date))
    .replace(/\{\{dateJa\}\}/g, formatDateJa(date));
}

export async function getDailyNotePath(date: Date): Promise<string> {
  return `daily/${formatDate(date)}.md`;
}

export async function ensureDailyNote(date: Date) {
  const path = await getDailyNotePath(date);

  if (await noteExists(path)) {
    return readNote(path);
  }

  const template = await getTemplate();
  const content = applyTemplate(template, date);

  // Parse out frontmatter and body from template
  const fmMatch = content.match(/^---\n([\s\S]*?)\n---\n([\s\S]*)$/);

  if (fmMatch) {
    // Template has frontmatter - create note with the full raw content
    const { writeFile, mkdir } = await import("node:fs/promises");
    const { resolve, dirname } = await import("node:path");
    const { ulid } = await import("ulid");
    const vaultPath = await getVaultPath();
    const fullPath = resolve(vaultPath, path);

    // Inject id and dates into frontmatter
    const now = new Date().toISOString();
    let fm = fmMatch[1];
    if (!fm.includes("id:")) fm = `id: "${ulid()}"\n${fm}`;
    if (!fm.includes("created:")) fm = `${fm}\ncreated: "${now}"`;
    if (!fm.includes("modified:")) fm = `${fm}\nmodified: "${now}"`;

    const rawContent = `---\n${fm}\n---\n${fmMatch[2]}`;

    await mkdir(dirname(fullPath), { recursive: true });
    await writeFile(fullPath, rawContent, "utf-8");

    return readNote(path);
  }

  // Fallback: simple creation
  return createNote(path, formatDate(date));
}

export function getRecentDailyDates(count: number = 7): Date[] {
  const dates: Date[] = [];
  const today = new Date();

  for (let i = 0; i < count; i++) {
    const date = new Date(today);
    date.setDate(today.getDate() - i);
    dates.push(date);
  }

  return dates;
}

File: internal/service/daily.go
package service

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type DailyService struct {
	vault *Vault
}

func NewDaily(vault *Vault) *DailyService {
	return &DailyService{vault: vault}
}

func DailyNotePath(date time.Time) string {
	return "daily/" + date.Format("2006-01-02") + ".md"
}

func (d *DailyService) EnsureDailyNote(date time.Time) (*VaultNote, error) {
	path := DailyNotePath(date)

	if d.vault.NoteExists(path) {
		return d.vault.ReadNote(path)
	}

	template := d.getTemplate()
	content := applyTemplate(template, date)

	// Parse frontmatter from template
	fm, body := parseFrontmatter([]byte(content))
	if fm.ID == "" {
		fm.ID = newULID()
	}
	now := time.Now().UTC().Format(time.RFC3339)
	if fm.Created == "" {
		fm.Created = now
	}
	if fm.Modified == "" {
		fm.Modified = now
	}
	if fm.Title == "" {
		fm.Title = date.Format("2006-01-02")
	}
	if fm.Tags == nil {
		fm.Tags = []string{"daily"}
	}

	return d.vault.WriteNote(path, body, fm)
}

func (d *DailyService) getTemplate() string {
	configPath := filepath.Join(d.vault.Path(), ".prometheus", "config.json")
	data, err := os.ReadFile(configPath)
	if err == nil {
		var config struct {
			DailyNoteTemplate string `json:"dailyNoteTemplate"`
		}
		if json.Unmarshal(data, &config) == nil && config.DailyNoteTemplate != "" {
			return config.DailyNoteTemplate
		}
	}

	return `---
title: "{{date}}"
tags: [daily]
---

# {{dateJa}}

## タスク

- [ ]

## メモ

`
}

func applyTemplate(template string, date time.Time) string {
	dateStr := date.Format("2006-01-02")

	// Japanese date format
	weekdays := []string{"日曜日", "月曜日", "火曜日", "水曜日", "木曜日", "金曜日", "土曜日"}
	dateJa := date.Format("2006年1月2日") + weekdays[date.Weekday()]

	result := strings.ReplaceAll(template, "{{date}}", dateStr)
	result = strings.ReplaceAll(result, "{{dateJa}}", dateJa)
	return result
}

func RecentDailyDates(count int) []time.Time {
	dates := make([]time.Time, count)
	today := time.Now()
	for i := 0; i < count; i++ {
		dates[i] = today.AddDate(0, 0, -i)
	}
	return dates
}
